plugins/nitrokey/fu-plugin-nitrokey.c

```diff
@@ -0,0 +1,365 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+
+#include <appstream-glib.h>
+#include <string.h>
+
+#include "fu-plugin.h"
+#include "fu-plugin-vfuncs.h"
+
+#define NITROKEY_TRANSACTION_TIMEOUT		1000 /* ms */
+#define NITROKEY_NR_RETRIES			5
+
+#define GET_DEVICE_STATUS			(0x20 + 14)
+
+typedef struct __attribute__((packed)) {
+	guint8		command;
+	guint8		payload[59];
+	guint32		crc;
+} NitrokeyHidRequest;
+
+typedef struct __attribute__((packed)) {
+	guint8		_padding; /* always zero */
+	guint8		device_status;
+	guint32		last_command_crc;
+	guint8		last_command_status;
+	guint8		payload[53];
+	guint32		crc;
+} NitrokeyHidResponse;
+
+/* based from libnitrokey/stick20_commands.h */
+typedef struct __attribute__((packed)) {
+	guint8		_padding[24];
+	guint8		SendCounter;
+	guint8		SendDataType;
+	guint8		FollowBytesFlag;
+	guint8		SendSize;
+	guint16		MagicNumber_StickConfig;
+	guint8		ReadWriteFlagUncryptedVolume;
+	guint8		ReadWriteFlagCryptedVolume;
+	guint8		VersionReserved1;
+	guint8		VersionMinor;
+	guint8		VersionReserved2;
+	guint8		VersionMajor;
+	guint8		ReadWriteFlagHiddenVolume;
+	guint8		FirmwareLocked;
+	guint8		NewSDCardFound;
+	guint8		SDFillWithRandomChars;
+	guint32		ActiveSD_CardID;
+	guint8		VolumeActiceFlag;
+	guint8		NewSmartCardFound;
+	guint8		UserPwRetryCount;
+	guint8		AdminPwRetryCount;
+	guint32		ActiveSmartCardID;
+	guint8		StickKeysNotInitiated;
+} NitrokeyGetDeviceStatusPayload;
+
+static guint32
+_stm_crc32_mutate (guint32 crc, guint32 data)
+{
+	crc = crc ^ data;
+	for (guint i = 0; i < 32; i++) {
+		if (crc & 0x80000000) {
+			/* polynomial used in STM32 */
+			crc = (crc << 1) ^ 0x04C11DB7;
+		} else {
+			crc = (crc << 1);
+		}
+	}
+	return crc;
+}
+
+static guint32
+_stm_crc32 (const guint8 *data, gsize size)
+{
+	guint32 crc = 0xffffffff;
+	const guint32 *pend = (const guint32 *) (data + size);
+	for (const guint32 *p = (const guint32 *) data; p < pend; p++)
+		crc = _stm_crc32_mutate (crc, *p);
+	return crc;
+}
+
+static void
+_dump_to_console (const gchar *title, const guint8 *buf, gsize buf_sz)
+{
+	if (g_getenv ("NITROKEY_DEBUG") == NULL)
+		return;
+	g_debug ("%s", title);
+	for (guint i = 0; i < 64; i++)
+		g_debug ("%u=0x%02x", i, buf[i]);
+}
+
+static gboolean
+nitrokey_execute_cmd (GUsbDevice *usb_device, guint8 command,
+		      const guint8 *buf_in, gsize buf_in_sz,
+		      guint8 *buf_out, gsize buf_out_sz,
+		      GCancellable *cancellable, GError **error)
+{
+	NitrokeyHidRequest req;
+	NitrokeyHidResponse res;
+	gboolean ret;
+	gsize actual_len = 0;
+	guint32 crc_tmp;
+
+	g_return_val_if_fail (buf_in_sz <= 59, FALSE);
+	g_return_val_if_fail (buf_out_sz <= 54, FALSE);
```

Comment:
these are kind of magic numbers, maybe a #define for them?

Reply:
Agreed, and this uncovered a bug as 54!=53... Fixing.